File: src/action/HttpRequest.php
<?php


namespace Ething\Action;


use \Ething\Url;
use \Ething\Proxy;
use \Ething\Request;
use \Ething\Response;
use \Ething\StreamBuffer;


class HttpRequest extends Action {
	
	static public function validate(array &$attributes, array $context){
		
		$attributes = array_merge(
			array( 
				'url' => null,
				'method' => null,
				'auth' => null,
				'headers' => array(),
				'body' => null,
				'output' => null
			),
			$attributes
		);
		
		foreach(array_keys($attributes) as $key){
			
			switch($key){
				
				case 'url':
					
					if(!Url::validate($attributes[$key]))
						throw new \Exception("field '{$key}' is not a valid url.");
					
					break;
				
				case 'method':
					
					if(!is_string($attributes[$key]) || !in_array(strtolower($attributes[$key]), Request::$validMethodes))
						throw new \Exception("field '{$key}' is not a valid method.");
					
					break;
					
				case 'headers':
					
					if(is_null($attributes[$key]))
						$attributes[$key] = array();
					
					if(is_object($attributes[$key]))
						$attributes[$key] = (array)$attributes[$key];
					
					if(!is_array($attributes[$key]))
						throw new \Exception("field '{$key}' is invalid.");
					
					foreach($attributes[$key] as $key => $value){
						if(!is_string($key) || !is_string($value) || empty($value))
							throw new \Exception("field '{$key}.{$key}' must be a non empty string.");
					}
					
					break;
				
				case 'auth':
					
					$value = &$attributes[$key];
					
					if(is_object($value))
						$value = (array)$value;
					
					if(is_array($value) && !empty($value['type']) && !empty($value['user']) && !empty($value['password']) && is_string($value['type']) && is_string($value['user']) && is_string($value['password'])){
						
						$authMode = strtolower($value['type']);
						$authUser = $value['user'];
						$authPwd = $value['password'];
						
						if(in_array($authMode, Request::$validAuthModes)){
							$value = array(
								'type' => $authMode,
								'user' => $authUser,
								'password' => $authPwd
							);
						} else {
							throw new \Exception("field '{$key}' : invalid auth mode.");
						}
					}
					else if(!is_null($value))
						throw new \Exception("field '{$key}' is invalid.");
					
					break;
					
				case 'body':
					
					$value = &$attributes[$key];
					
					if(is_string($value)){
						$value = array(
							'type' => 'plain',
							'value' => $value
						);
					}
					
					if(is_object($value))
						$value = (array)$value;
					
					if(is_array($value) && !empty($value['type']) && !empty($value['value']) && is_string($value['type']) && is_string($value['value'])){
						
						$bodyMode = strtolower($value['type']);
						$bodyContent = $value['value'];
						
						$value = array(
							'type' => $bodyMode,
							'value' => $bodyContent
						);
						
						switch($bodyMode){
							case 'plain':
								
								if(!is_string($bodyContent) || strlen($bodyContent) > 1024)
									throw new \Exception("field '{$key}.value' must be a string with a maximum size of 1024 characters.");
								
								break;
								
							case 'binary':
								
								if(!is_string($bodyContent) || strlen($bodyContent) > 65536 || base64_decode($bodyContent) === false)
									throw new \Exception("field '{$key}.value' must be a base64 encoded string with a maximum size of 64 kbytes.");
								
								break;
							case 'resource': 
								
								if(!is_string($bodyContent))
									throw new \Exception("field '{$key}.value' must be an id.");
								
								$resource = $context['ething']->get($bodyContent);
								if(!$resource)
									throw new \Exception("the resource with id '{$bodyContent}' does not exist.");
								
								if(!in_array($resource->type(), array('Table','File')))
									throw new \Exception("the resource with id '{$bodyContent}' must be a Table or a File.");
								
								break;
							
							default:
								throw new \Exception("field '{$key}.type' : unknown body type.");
								
						}
						
					}
					else if(!is_null($value))
						throw new \Exception("field '{$key}' is invalid.");
					
					break;
				
				case 'output':
					
					if(!((is_string($attributes[$key]) || !empty($attributes[$key])) || is_null($attributes[$key])))
						throw new \Exception("field '{$key}' must be a non empty string or unset.");
					
					break;
					
				default:
					throw new \Exception("field '{$key}' unknown.");
			}
			
		}
		
		return true; 
	}
	
	
	protected function call(\Ething\Event\Signal $signal){
		
		$body = null;
		$contentType = null;
		
		if(!is_null($this->body)){
			switch($this->body['type']){
				case 'plain':
					$body = $this->body['value'];
					$contentType = 'text/plain';
					break;
				
				case 'binary':
					$body = base64_decode($this->body['value']);
					$contentType = 'application/octet-stream';
					break;
				
				case 'resource': 
					
					$src = $this->ething()->get( $this->body['value'] );
					if(!$src)
						throw new \Ething\InvalidRuleException("The resource #{$this->body['value']} does not exist any more");
					
					if($src instanceof \Ething\File){
						$body = $src->read();
						$contentType = $src->mime;
					}
					else if($src instanceof \Ething\Table){
						$body = \json_endode( $src->select(), JSON_PRETTY_PRINT|JSON_UNESCAPED_SLASHES);
						$contentType = 'application/json';
					}
					else if($src instanceof \Ething\App){
						$body = $src->readScript();
						$contentType = 'text/html';
					}
					
					break;
					
			}
		}
		
		
		
		$request = new Request($this->url, $this->method, $this->headers, $body);
		
		if($contentType && !$request->contentType())
			$request->addHeader('content-type', $contentType);
		
		$authMode = isset($this->auth['type']) ? $this->auth['type'] : null;
		$authUser = isset($this->auth['user']) ? $this->auth['user'] : null;
		$authPwd = isset($this->auth['password']) ? $this->auth['password'] : null;
		
		
		// send the request
		$proxy = new Proxy($this->ething());
		$buffer = new StreamBuffer();
		
		$response = $proxy->request($request, $buffer, $authUser, $authPwd, $authMode, array(
			CURLOPT_TIMEOUT => 20 // in secondes
		));
		
		if(!$response){
			throw new \Exception("unable to connect to the host");
		}
		
		if(!$response->isSuccessful()){
			throw new \Exception( sprintf("status: %d", $response->getStatus()) );
		}
		
		if(!is_null($this->output) && $buffer->length()){	
			//create a new file
			if($file = $this->ething()->create('File', array( 'name' => $this->output ))){
				$file->write( $buffer->data() ); // put the content in it !
			}
		}
		
	}
	
}

File: ething/rule/action/HttpRequest.py
# coding: utf-8
from future.utils import string_types
from .. import Action
from .. import InvalidRuleException
import requests
import base64
import json

class HttpRequest(Action):
    
    @staticmethod
    def validate (attributes, context):
        
        attributes.setdefault('url', None)
        attributes.setdefault('method', 'GET')
        attributes.setdefault('auth', None)
        attributes.setdefault('headers', None)
        attributes.setdefault('body', None)
        attributes.setdefault('output', None)
        
        for key in attributes:
            
            value = attributes[key]
            
            if key == 'url':
                
                if not( isinstance(value, string_types) and len(value)>0 ):
                    raise Exception("%s: must be a non empty string." % key)
            
            elif key == 'method':
                
                if not( isinstance(value, string_types) and value.lower() in ['get', 'post', 'delete', 'put', 'patch'] ):
                    raise Exception("%s: invalid" % key)
            
            elif key == 'headers':
                
                if value is not None:
                    
                    if isinstance(value, dict):
                        for k in value:
                            if not isinstance(value[k], string_types):
                                raise Exception("%s: invalid" % key)
                    else :
                        raise Exception("%s: invalid" % key)
                
            elif key == 'auth':
                
                if value is not None:
                    if not isinstance(value, dict) or 'type' not in value or not isinstance(value['type'], string_types) or 'user' not in value or not isinstance(value['user'], string_types) or 'password' not in value or not isinstance(value['password'], string_types) :
                        raise Exception("%s: invalid" % key)
                    
                    if value['type'].lower() not in ['basic', 'digest']:
                        raise Exception("%s: invalid" % key)
                    
                    o = {
                        'type': value['type'].lower(),
                        'user': value['user'],
                        'password': value['password']
                    }
                    
                    attributes[key] = o
            
            elif key == 'body':
                
                if value is not None:
                    
                    if isinstance(value, string_types):
                        value = {
                            'type' : 'plain',
                            'value' : value
                        }
                    
                    if not isinstance(value, dict) or 'type' not in value or not isinstance(value['type'], string_types) or 'value' not in value or not isinstance(value['value'], string_types):
                        raise Exception("%s: invalid" % key)
                    
                    if value['type'].lower() not in ['plain','binary','resource']:
                        raise Exception("%s: invalid" % key)
                    
                    o = {
                        'type': value['type'].lower(),
                        'value': value['value']
                    }
                    
                    if o['type'] == 'resource':
                        
                        resource = context['ething'].get(value['value'])
                        if not resource:
                            raise Exception("%s: the resource with id '%s' does not exist." % (key,value['value']))
                        
                        if resource.type not in ['Table','File']:
                            raise Exception("%s: the resource with id '%s' must be a Table or a File." % (key,str(resource)))
                    
                    attributes[key] = o
            
            elif key == 'output':
                
                if value is not None:
                    if not( isinstance(value, string_types) and len(value)>0 ):
                        raise Exception("%s: must be resource filename." % key)
            
            else:
                raise Exception("%s: invalid" % key)
        
        return True
    
    
    
    def call(self, signal):
        body = None
        contentType = None
        
        if self['body']:
            body = self['body']
            
            if body['type'] == 'plain':
                body = body['value']
                contentType = 'text/plain'
            
            elif body['type'] ==  'binary':
                body = base64.b64decode(body['value'])
                contentType = 'application/octet-stream'
            
            elif body['type'] ==  'resource': 
                
                src = self.ething.get( body['value'] )
                if not src:
                    raise InvalidRuleException("The resource id='%s' does not exist any more" % body['value'])
                
                if src.type == 'File':
                    body = src.read()
                    contentType  = src.mime
                elif src.type == 'Table':
                    body = json.dumps(src.select(), indent = 4)
                    contentType  = 'application/json'
        
        auth = None
        
        if self['auth']:
            if self['auth']['type'] == 'basic':
                auth = requests.auth.HTTPBasicAuth(self['auth']['user'], self['auth']['password'])
            elif self['auth']['type'] == 'digest':
                auth = requests.auth.HTTPDigestAuth(self['auth']['user'], self['auth']['password'])
        
        s = requests.Session()
        if auth is not None:
            s.auth = auth
        
        #s.config['keep_alive'] = False # cf. https://stackoverflow.com/questions/10115126/python-requests-close-http-connection
        
        headers = self['headers']
        
        if contentType is not None:
            headers['content-type'] = contentType
        
        req = requests.Request(self['method'].upper(), self['url'], headers)
        prepped = s.prepare_request(req)
        
        if body:
            prepped.body = body
        
        resp = s.send(prepped,
            timeout=(6.05, 15)
        )
        
        if self['output']:
            
            text = resp.text
            
            if text:
                outf = self.ething.create('File',{
                    'name': self['output']
                })
                if outf:
                    outf.write(str(text))
